File: python/src/sendsoon_mcp/tools/markitdown_convert.py
"""markitdown_convert MCP tool."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from sendsoon_mcp.markitdown_file import load_markitdown_file_from_path

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP

    from sendsoon_mcp.client import SendSoonClient

TOOL_DESCRIPTION = (
    "Convert a local file (pdf, docx, pptx, xlsx, images, audio, csv, json, html, zip, "
    "epub, txt, etc.) to Markdown text via SendSoon. Provide file_path; the file name is "
    "detected automatically (max 10 MB)."
)


def register(mcp: FastMCP, client: SendSoonClient) -> None:
    @mcp.tool(name="markitdown_convert", description=TOOL_DESCRIPTION)
    async def markitdown_convert(file_path: str) -> dict[str, Any]:
        """Convert a local file to Markdown via SendSoon."""
        request, error = load_markitdown_file_from_path(file_path)
        if error:
            return {"success": False, "error": error.to_dict()}
        assert request is not None
        return await client.markitdown_convert(request)

File: npm/src/tools/markitdown_convert.ts
import {
  loadMarkitdownFileFromPath,
  markitdownFailureResult,
  type MarkitdownConvertResult,
  type SendSoonClient,
} from '../core/index.js';
import * as z from 'zod/v4';
import { formatToolResult } from './format.js';

const markitdownConvertInputSchema = {
  file_path: z
    .string()
    .trim()
    .min(1)
    .describe('Local path to the file to convert. The file name is detected automatically.'),
};

const markitdownConvertOutputSchema = {
  success: z.boolean(),
  filename: z.string().optional(),
  markdown: z.string().optional(),
  error: z
    .object({
      code: z.string(),
      message: z.string(),
      retryable: z.boolean(),
    })
    .optional(),
};

export type MarkitdownConvertInput = {
  file_path: string;
};

export type MarkitdownConvertOutput = MarkitdownConvertResult;

export const markitdownConvertToolDefinition = {
  name: 'markitdown_convert' as const,
  config: {
    title: '文件转 Markdown',
    description:
      'Convert a local file (pdf, docx, pptx, xlsx, images, audio, csv, json, html, zip, epub, txt, etc.) to Markdown text via SendSoon. Provide file_path; the file name is detected automatically (max 10 MB).',
    inputSchema: markitdownConvertInputSchema,
    outputSchema: markitdownConvertOutputSchema,
  },
  createHandler(client: SendSoonClient) {
    return async (input: MarkitdownConvertInput) => {
      const loaded = await loadMarkitdownFileFromPath(input.file_path);
      if ('error' in loaded) {
        return formatToolResult(markitdownFailureResult(loaded.error));
      }

      const result = await client.markitdownConvert(loaded.request);
      return formatToolResult(result);
    };
  },
};
